sdk/core/src/error/mod.rs

```diff
@@ -0,0 +1,371 @@
+use std::borrow::Cow;
+use std::fmt::{Debug, Display};
+
+mod azure_core_errors;
+mod hyperium_http;
+mod macros;
+
+/// A convience alias for `Result` where the error type is hard coded to `Error`
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// The kind of error
+///
+/// The classification of error is intentionally fairly coarse.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum ErrorKind {
+    /// An HTTP status code that was not expected
+    HttpResponse { status: u16 },
```

Comment:
Per our [guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#handling-errors) and discussion Tuesday on Azure errors, I think we should include both the `status` (HTTP status code) as well as the `code`  (or perhaps `error_code` would be more descriptive, and what we use for .NET's `RequestFailedException`). The body I know is captured as additional data for the `Error` itself, but the goal is that decisions can be made on both the status and the error code in the standard Azure error.

Reply:
This would need a bit more thought. The `ErrorKind` is currently `Copy` because it's meant to be very basic data. Adding a `String` `error_code` would change this, but perhaps that's ok?

cc @yoshuawuyts

Reply:
Couldn't we just copy the string, or would that be considered too expensive for an error? They'd typically (always?) be short - a dozen or so characters, typically.

But if this isn't the right place, we should still probably have it explicitly available as context on the containing `Error`. Still, given we're putting the `status` here I think it makes sense to have the `error_code` here as well.

Reply:
What you're describing is `Clone` in Rust. `Copy` is essentially a memcopy and changes types from moving by default to being copied. This excludes all types that have any pointers (including string) because otherwise you could end up with aliased owning pointers which is not allowed in Rust.

Essentially, yes we can do what you're suggesting, but that means exactly what I said above, we remove `Copy` and the type just becomes `Clone`.